elements/pl-string-input/pl-string-input.py

```diff
@@ -0,0 +1,228 @@
+        data['submitted_answers'][name] = None
+        return
+
+    try:
+        # check if answer is left blank
+        if not a_sub:
+            if not allow_blank:
+                raise ValueError('invalid submitted answer (left blank)')
+        data['submitted_answers'][name] = pl.to_json(a_sub)
+    except Exception:
+        data['format_errors'][name] = 'Invalid format. The submitted answer was left blank.'
+        data['submitted_answers'][name] = None
+
+
+def grade(element_html, element_index, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, 'answers_name')
+
+    # Get weight
+    weight = pl.get_integer_attrib(element, 'weight', 1)
+
+    # Get remove_between_spaces option
+    remove_between_spaces = pl.get_string_attrib(element, 'remove_between_spaces', False)
+
+    # Get remove_leading_trailing option
+    remove_leading_trailing = pl.get_string_attrib(element, 'remove_leading_trailing', False)
+
+    # Get true answer (if it does not exist, create no grade - leave it
+    # up to the question code)
+    a_tru = pl.from_json(data['correct_answers'].get(name, None))
+    if a_tru is None:
+        return
+
+    # Get submitted answer (if it does not exist, score is zero)
+    a_sub = data['submitted_answers'].get(name, None)
+    if a_sub is None:
+        data['partial_scores'][name] = {'score': 0, 'weight': weight}
+        return
+    # If submitted answer is in a format generated by pl.to_json, convert it
+    # back to a standard type (otherwise, do nothing)
+    a_sub = pl.from_json(a_sub)
+
+    # Remove the leading and trailing characters
+    if (remove_leading_trailing):
+        a_sub = a_sub.strip()
+
+    # Remove the blank spaces between characters
+    if (remove_between_spaces):
+        a_sub = a_sub.replace(' ', '')
+
+    if a_tru == a_sub:
+        data['partial_scores'][name] = {'score': 1, 'weight': weight}
+    else:
+        data['partial_scores'][name] = {'score': 0, 'weight': weight}
+
+
+def test(element_html, element_index, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, 'answers_name')
+    weight = pl.get_integer_attrib(element, 'weight', 1)
+
+    # Get correct answer
+    a_tru = data['correct_answers'][name]
+
+    # If correct answer is in a format generated by pl.to_json, convert it
+    # back to a standard type (otherwise, do nothing)
+    a_tru = pl.from_json(a_tru)
+
+    result = random.choices(['correct', 'incorrect'], [5, 5])[0]
+
+    if result == 'correct':
+        data['raw_submitted_answers'][name] = a_tru
+        data['partial_scores'][name] = {'score': 1, 'weight': weight}
+    elif result == 'incorrect':
+        data['raw_submitted_answers'][name] = a_tru + str((random.randint(1, 11) * random.choice([-1, 1])))
+        data['partial_scores'][name] = {'score': 0, 'weight': weight}
+    # No test for 'invalid' implemented at this time
```

Comment:
Can you add an invalid test? Returning a blank submission would test invalid, I believe?